fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/ReorderJoin.java

```diff
@@ -0,0 +1,201 @@
+                return join;
+            } else {
+                return new LogicalFilter(ExpressionUtils.and(nonJoinConditions), join);
+            }
+        } else {
+            Plan left = joinInputs.get(0);
+            List<Plan> candidate = joinInputs.subList(1, joinInputs.size());
+
+            List<Slot> leftOutput = left.getOutput();
+            Optional<Plan> rightOpt = candidate.stream().filter(right -> {
+                List<Slot> rightOutput = right.getOutput();
+
+                Set<Slot> joinOutput = getJoinOutput(left, right);
+                Optional<Expression> joinCond = conjuncts.stream()
+                        .filter(expr -> {
+                            Set<Slot> exprInputSlots = SlotExtractor.extractSlot(expr);
+                            if (exprInputSlots.isEmpty()) {
+                                return false;
+                            }
+
+                            if (new HashSet<>(leftOutput).containsAll(exprInputSlots)) {
+                                return false;
+                            }
+
+                            if (new HashSet<>(rightOutput).containsAll(exprInputSlots)) {
+                                return false;
+                            }
+
+                            return joinOutput.containsAll(exprInputSlots);
+                        }).findFirst();
+                return joinCond.isPresent();
+            }).findFirst();
+
+            Plan right = rightOpt.orElseGet(() -> candidate.get(1));
+            Set<Slot> joinOutput = getJoinOutput(left, right);
+            Map<Boolean, List<Expression>> split = splitConjuncts(conjuncts, joinOutput);
+            List<Expression> joinConditions = split.get(true);
+            List<Expression> nonJoinConditions = split.get(false);
+
+            Optional<Expression> cond;
+            if (joinConditions.isEmpty()) {
+                cond = Optional.empty();
+            } else {
+                cond = Optional.of(ExpressionUtils.and(joinConditions));
+            }
+
+            LogicalJoin join = new LogicalJoin(JoinType.INNER_JOIN, cond, left, right);
+
+            List<Plan> newInputs = new ArrayList<>();
+            newInputs.add(join);
+            newInputs.addAll(candidate.stream().filter(plan -> !right.equals(plan)).collect(Collectors.toList()));
+            return reorderJoins(newInputs, nonJoinConditions);
+        }
+    }
+
+    private Set<Slot> getJoinOutput(Plan left, Plan right) {
+        HashSet<Slot> joinOutput = new HashSet<>();
+        joinOutput.addAll(left.getOutput());
+        joinOutput.addAll(right.getOutput());
+        return joinOutput;
+    }
+
+    private Map<Boolean, List<Expression>> splitConjuncts(List<Expression> conjuncts, Set<Slot> slots) {
+        return conjuncts.stream().collect(Collectors.partitioningBy(
+                // TODO: support non equal to conditions.
+                expr -> expr instanceof EqualTo && slots.containsAll(SlotExtractor.extractSlot(expr))));
```

Comment:
maybe we need to check left child and right child slots on the different side of the equation

Reply:
Actually, this doesn't make a big difference. 
This rule push predicates with slots from one input into the join. 
`PushPredicateThroughJoin` could push this kind of predicates through join, both in join conditions or filter predicates.